### PR TITLE
implement ALPN support

### PR DESCRIPTION
**Pull request by ekoby**

### CODE REVIEW

File: src/engine_openssl.c [hunk 1]
```diff
@@ -295,6 +305,9 @@ static int tls_verify_signature(void *cert, enum hash_algo md, const char* data,
 
 static void tls_free_ctx(tls_context *ctx) {
     struct openssl_ctx *c = ctx->ctx;
+    if (c->alpn_protocols) {
+        free(c->alpn_protocols);
```
Comment (dovholuknf) [inline src/engine_openssl.c:309]: shouldn't this free each entry like how `mbedtls_set_alpn_protocols` does? also - why not make this a function and stay DRY? same comment below in `tls_set_alpn_protocols`

File: src/engine_openssl.c [hunk 2]
```diff
@@ -331,6 +347,30 @@ static void tls_free_cert(tls_cert *cert) {
...
+    }
+
+    size_t protolen = 0;
+    for (int i=0; i < len; i++) {
+        protolen = strlen(protos[i]) + 1;
+    }
+
+    c->alpn_protocols = malloc(protolen + 1);
```
Comment (dovholuknf) [inline src/engine_openssl.c:362]: shouldn't this be calloc not malloc?

Comment (ekoby) [reply src/engine_openssl.c:364]: we assign all bytes

File: src/engine_openssl.c (hunk 1 above)
Comment (ekoby) [reply src/engine_openssl.c:311]: openssl uses different structure to configure ALPN

File: src/engine_openssl.c [hunk 3]
```diff
@@ -331,6 +349,30 @@ static void tls_free_cert(tls_cert *cert) {
...
+static void tls_set_alpn_protocols(void *ctx, const char **protos, int len) {
+    struct openssl_ctx *c = ctx;
+
+    if (c->alpn_protocols) {
+        free(c->alpn_protocols);
+    }
+
+    size_t protolen = 0;
+    for (int i=0; i < len; i++) {
+        protolen = strlen(protos[i]) + 1;
```
Comment (dovholuknf) [inline src/engine_openssl.c:361]: looking at this again - why do this in a loop (multip assignments of protolen)? is this meant to be += here?

Comment (ekoby) [reply src/engine_openssl.c]: correct

**Review by dovholuknf — APPROVED** [review]

**Pull request merged**